compiler/types.nim

```diff
@@ -587,79 +606,106 @@ proc typeToString(typ: PType, prefer: TPreferedDesc = preferName): string =
-  of tyRange:
-    assert(t.n != nil)        # range directly given:
-    assert(t.n.kind == nkRange)
-    result = getOrdValue(t.n.sons[0])
-  of tyInt:
-    if platform.intSize == 4: result = - (2147483646) - 2
-    else: result = 0x8000000000000000'i64
-  of tyInt8: result = - 128
-  of tyInt16: result = - 32768
-  of tyInt32: result = - 2147483646 - 2
-  of tyInt64: result = 0x8000000000000000'i64
-  of tyUInt..tyUInt64: result = 0
-  of tyEnum:
-    # if basetype <> nil then return firstOrd of basetype
-    if sonsLen(t) > 0 and t.sons[0] != nil:
-      result = firstOrd(t.sons[0])
+proc firstValue*[T:BiggestInt|BiggestFloat](t: PType): T =
+  case t.kind:
+    of tyVar: result = firstValue[T](t.sons[0])
+    of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias:
+      result = firstValue[T](lastSon(t))
+    of tyRange:
+      assert(t.n != nil)        # range directly given:
+      assert(t.n.kind == nkRange)
+      result = getValue[T](t.n.sons[0])
     else:
-      assert(t.n.sons[0].kind == nkSym)
-      result = t.n.sons[0].sym.position
-  of tyGenericInst, tyDistinct, tyTypeDesc, tyAlias:
-    result = firstOrd(lastSon(t))
-  of tyOrdinal:
-    if t.len > 0: result = firstOrd(lastSon(t))
-    else: internalError("invalid kind for first(" & $t.kind & ')')
-  else:
-    internalError("invalid kind for first(" & $t.kind & ')')
-    result = 0
-
-proc lastOrd*(t: PType; fixedUnsigned = false): BiggestInt =
+      when T is BiggestInt:
+        case t.kind:
+        of tyBool, tyChar, tySequence, tyOpenArray, tyString, tyVarargs, tyProxy:
+          result = 0
+        of tySet, tyVar: result = firstValue[T](t.sons[0])
+        of tyArray: result = firstValue[T](t.sons[0])
+        of tyInt:
+          if platform.intSize == 4: result = - (2147483646) - 2
+          else: result = 0x8000000000000000'i64
+        of tyInt8: result = - 128
+        of tyInt16: result = - 32768
+        of tyInt32: result = - 2147483646 - 2
+        of tyInt64: result = 0x8000000000000000'i64
+        of tyUInt..tyUInt64: result = 0
+        of tyEnum:
+          # if basetype <> nil then return firstOrd of basetype
+          if sonsLen(t) > 0 and t.sons[0] != nil:
+            result = firstValue[T](t.sons[0])
+          else:
+            assert(t.n.sons[0].kind == nkSym)
+            result = t.n.sons[0].sym.position
+        of tyOrdinal:
+          if t.len > 0: result = firstValue[T](lastSon(t))
+          else: internalError("invalid kind for first(" & $t.kind & ')')
+        else:
+          internalError("invalid kind for first(" & $t.kind & ')')
+          result = 0
+
+      elif T is BiggestFloat:
+        case t.kind:
+          of tyFloat..tyFloat128: result = -Inf
+          else:
+            internalError("invalid kind for first(" & $t.kind & ')')
+            result = NaN
+
+proc firstOrd*(t: PType): BiggestInt {.inline.} =
+  firstValue[BiggestInt](t)
+
+proc lastValue*[T:BiggestInt|BiggestFloat](t: PType; fixedUnsigned = false): T =
```

Comment:
I dislike this style. It's not a "generic", it's 2-3 different implementations which happen to use the same name. And they are not generic.

Reply:
done